File: DevExpress-Demo/CS/RichEditDemo.Wpf/Modules/MailMerge/DBUriStreamProvider.cs
using System.Collections.Generic;
using System.IO;
using System.Linq;
using DevExpress.DemoData.Models;
using DevExpress.Office.Services;

namespace RichEditDemo {
    public class DBUriStreamProvider : IUriStreamProvider {
        static readonly string prefix = "dbimg://";

        readonly IList<Employee> employees;

        public DBUriStreamProvider(IList<Employee> employees) {
            this.employees = employees;
        }

        Stream IUriStreamProvider.GetStream(string uri) {
            uri = uri.Trim();
            if (!uri.StartsWith(prefix))
                return null;
            string strId = uri.Substring(prefix.Length).Trim();
            int id;
            if (!int.TryParse(strId, out id))
                return null;
            byte[] bytes = employees.First(employee => employee.EmployeeID == id).Photo;
            if (bytes == null)
                return null;
            return new MemoryStream(bytes);
        }
    }
}

File: DevExpress-Demo/VB/RichEditDemo.Wpf/Modules/MailMerge/DBUriStreamProvider.vb
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
Imports DevExpress.DemoData.Models
Imports DevExpress.Office.Services

Namespace RichEditDemo

    Public Class DBUriStreamProvider
        Implements IUriStreamProvider

        Private Shared ReadOnly prefix As String = "dbimg://"

        Private ReadOnly employees As IList(Of Employee)

        Public Sub New(ByVal employees As IList(Of Employee))
            Me.employees = employees
        End Sub

        Private Function GetStream(ByVal uri As String) As Stream Implements IUriStreamProvider.GetStream
            uri = uri.Trim()
            If Not uri.StartsWith(prefix) Then Return Nothing
            Dim strId As String = uri.Substring(prefix.Length).Trim()
            Dim id As Integer
            If Not Integer.TryParse(strId, id) Then Return Nothing
            Dim bytes As Byte() = Enumerable.First(employees, Function(employee) employee.EmployeeID = id).Photo
            If bytes Is Nothing Then Return Nothing
            Return New MemoryStream(bytes)
        End Function
    End Class
End Namespace
